#379 expose 'samDesired' param to all methods

CamW opened the pull request:
Exposes samDesired as discussed in the issue. Includes tests which use HKEY_LOCAL_MACHINE instead of HKEY_CURRENT_USER. This is because HKEY_LOCAL_MACHINE is redirected whereas HKEY_CURRENT_USER is shared. https://docs.microsoft.com/en-us/windows/desktop/winprog64/shared-registry-keys#redirected-shared-and-reflected-keys-under-wow64 Using HKEY_LOCAL_MACHINE allows for tests which ensure that the extra samDesired flags are applied by performing operations on the same registry keys across the 32 and 64 bit registry without conflict. Tests will only run on 64bit windows since WOW64 would be ignored on 32bit and break the tests.

matthiasblaesing commented on contrib/platform/src/com/sun/jna/platform/win32/Advapi32Util.java:
```diff
@@ -593,9 +609,28 @@ public static boolean registryKeyExists(HKEY root, String key) {
 	public static boolean registryValueExists(HKEY root, String key,
-			String value) {
+											  String value) {
+		return registryValueExists(root, key, value, WinNT.REG_NONE);
```
Please don't use `WinNT.REG_NONE` for this use-case. `WinNT.REG_NONE` is a type, not an access level. There is no "NONE" access level for the registry. I would just use 0 here. I would not invent a constant.

CamW commented:
I've switched out the bad WinNT.REG_NONEs with 0 as requested. The tests require admin rights because they're accessing HKLM. HKCU\Software isn't redirected as mentioned originally but HKCU\Software\Classes\CLSID is so I can use that instead. I've demonstrated the change in the testRegistryCreateDeleteKeySamExtra test in the latest commit. If you find that preferable, I can modify the others to match.

matthiasblaesing commented:
Thank you for the change - I think using the `HKCU\Software\Classes\CLSID` is better. Requiring administrator rights to run unittests should be avoided if possible (yes there are tests, that require it, but it should be an exception).

The unittest `testRegistryKeyExistsSamExtra` fails on appveyor. See: https://ci.appveyor.com/api/buildjobs/s6a0brgcwimsro9b/log (line 1916). Any idea why that fails and why the other checks succeed?

CamW commented:
In unit test `testRegistryKeyExistsSamExtra` I tried to use pre-existing registry entries. It's possible that those entries don't always exist. I've changed that test to a create-test-delete style test. 
I've also changed the rest of the samExtra tests to use the new `HKCU\Software\Classes\CLSID` path which means that Advapi32UtilTest no longer requires administrator rights.

matthiasblaesing commented:
Thank you - with the change the unittests run cleanly and also the appveyor build is happy.
